### PR TITLE
LSTM multi-stream solver

### PR DESCRIPTION
**Pull request by shurale-nkn**
PR adds new multi-stream solution for FWD LSTM.
* new solver
* new multi-stream mechanism in MIOpen handle
* changed time measurement algorithm in FWD LSTM at HIP runtime

gfx90a 104cu
rocm 5.3
FWD time comparison
layer | in_vec | h_size | seq | max_n | perf boost | develop time | PR#2034 time
-- | -- | -- | -- | -- | -- | -- | --
3 | 2048 | 1024 | 158 | 16 | 33% | 23.59 | 17.7
3 | 2048 | 1024 | 179 | 16 | 35% | 26.93 | 19.92
2 | 240 | 1024 | 315 | 16 | 36% | 29.85 | 21.87
2 | 240 | 1024 | 357 | 16 | 35% | 33.36 | 24.63
2 | 512 | 512 | 102 | 16 | 27% | 3.96 | 3.13
2 | 512 | 512 | 58 | 16 | 12% | 2.31 | 2.06
8 | 224 | 1000 | 224 | 8 | 35% | 75.14 | 55.5

### CODE REVIEW

**Comment by dmikushin** [comment]
Hi @shurale-nkn , congrats making something out of streams in LSTM! The code is very cryptic, it's hard to understand anything except that a pool of streams is created, and then parts of result are elegantly computed in parallel. The dependencies are resolved dynamically: each next steps checks whether it depends on another stream, and if it's the case - waits for it. Hard to understand anything further.

* Would you like maybe to make an inline comment to demonstrate the parallelization scheme?
* Would you like to define a higher-level set of functions like `submit()` or `wait_for_deps()`, etc that could be read without additional details?

**Comment by shurale-nkn** [comment]
> * Would you like maybe to make an inline comment to demonstrate the parallelization scheme?
> * Would you like to define a higher-level set of functions like `submit()` or `wait_for_deps()`, etc that could be read without additional details?

No

**Comment by dmikushin** [comment]
Could it be proven that cyclic dependencies between streams could not ever occur? I mean, it feels obvious that it couldn't, but could we be 100% sure?

**Review by junliume — APPROVED** [review]

**Review by DrizztDoUrden — APPROVED** [review]

File: src/hip/handlehip.cpp [hunk 1]
```diff
@@ -201,8 +205,42 @@ struct HandleImpl
...
+            rhandle_pool.push_back(std::move(r_ptr));
+        }
+#else
+        void add_stream(StreamPtr& s_ptr) { stream_pool.push_back(s_ptr); }
+#endif
+        //  stream_pool used as cache for parallel streams created by MIOpen.
+        StreamPtrPool stream_pool;
+    };
+
+    MultiStreamResourses* ms_resourse_ptr;
```
Comment (shurale-nkn) [inline src/hip/handlehip.cpp:240]: MultiStreamResourses -> MultiStreamResources
ms_resourse_ptr -> ms_resource_ptr

post merge renaming

**Comment by junliume** [comment]
Ping @JehandadKhan about to merge this PR with passing CI

**Comment by JehandadKhan** [comment]
@shurale-nkn Can you please share the driver commands corresponding to the above test cases so that we can verify the performance improvements resulting from this change.

cc @junliume  @aserio

**Comment by shurale-nkn** [comment]
> @shurale-nkn Can you please share the driver commands corresponding to the above test cases so that we can verify the performance improvements resulting from this change.
> 
> cc @junliume @aserio

@JehandadKhan 

MIOpenDriver rnn -m lstm -l 3 -W 2048 -H 1024 -n 16 -k 158  -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 3 -W 2048 -H 1024 -n 16 -k 179  -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 2 -W 240 -H 1024 -n 16 -k 315 -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 2 -W 240 -H 1024 -n 16 -k 357 -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 2 -W 512 -H 512 -n 16 -k 102 -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 2 -W 512 -H 512 -n 16 -k 58 -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0
MIOpenDriver rnn -m lstm -l 8 -W 224 -H 1000 -n 8 -k 224 -F 1 -r 0 -b 1 -p 0 -t 1 -w 1 -i 5 -V 0

**Comment by aserio** [comment]
Thanks!

**Pull request merged**